.github/workflows/wasm.yml

```diff
@@ -0,0 +1,129 @@
+  emsdk_version: 2.0.9
+
+jobs:
+  build-wasm:
+    name: "emscripten"
+    runs-on: ubuntu-latest
+    steps:
+
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          submodules: recursive
+
+      - name: Setup Emscripten toolchain
+        run: |
+            git clone --depth 1 https://github.com/emscripten-core/emsdk.git
+            (cd emsdk && ./emsdk install ${{ env.emsdk_version }} ccache-git-emscripten-64bit)
+            (cd emsdk && ./emsdk activate ${{ env.emsdk_version }} ccache-git-emscripten-64bit)
+            # mtime of this file is checked by ccache, we set it to avoid cache misses.
+            touch -m -d '1 Jan 2021 12:00' emsdk/.emscripten
+      - name: Setup Paths
+        run: |
+            # Following two variables are set by emsdk activated otherwise
+            echo "_EMCC_CCACHE=1" >> $GITHUB_ENV
+            echo "EM_CONFIG=$(realpath emsdk/.emscripten)" >> $GITHUB_ENV
+
+            # Configure GitHub Paths
+            echo "${{ github.workspace }}/emsdk" >> $GITHUB_PATH
+            echo "${{ github.workspace }}/emsdk/upstream/emscripten" >> $GITHUB_PATH
+            echo "${{ github.workspace }}/emsdk/node/14.18.2_64bit/bin" >> $GITHUB_PATH
+            echo "${{ github.workspace }}/emsdk/ccache/git-emscripten_64bit/bin" >> $GITHUB_PATH
+
+      - name: Generate ccache_vars for ccache based on machine
+        shell: bash
+        id: ccache_vars
+        run: |-
+          echo "::set-output name=hash::$(echo ${{ env.ccache_compilercheck }})"
+          echo "::set-output name=timestamp::$(date '+%Y-%m-%dT%H.%M.%S')"
+
+      - name: Verify Emscripten setup
+        run: |
+            emcc --version
+            emcmake cmake --version
+            emmake make --version
+
+      - name: Set ccache environment for emcc
+        run: |
+          echo "CCACHE_COMPILER_CHECK=${{ env.ccache_compilercheck }}" >> $GITHUB_ENV
+          echo "CCACHE_BASEDIR=${{ env.ccache_basedir }}" >> $GITHUB_ENV
+          echo "CCACHE_COMPRESS=${{ env.ccache_compress }}" >> $GITHUB_ENV
+          echo "CCACHE_COMPRESSLEVEL=${{ env.ccache_compresslevel }}" >> $GITHUB_ENV
+          echo "CCACHE_DIR=${{ env.ccache_dir }}" >> $GITHUB_ENV
+          echo "CCACHE_MAXSIZE=${{ env.ccache_maxsize }}" >> $GITHUB_ENV
+          # https://emscripten.org/docs/compiling/Building-Projects.html#using-a-compiler-wrapper
+          echo "EM_COMPILER_WRAPPER=ccache" >> $GITHUB_ENV
+          
+
+      - name: Cache-op for build-cache through ccache
+        uses: actions/cache@v2
+        with:
+          path: ${{ env.ccache_dir }}
+          key: ccache-${{ job.id }}-${{ steps.ccache_vars.outputs.hash }}-${{ github.ref }}-${{ steps.ccache_vars.outputs.timestamp }}
+          restore-keys: |-
+            ccache-${{ job.id }}-${{ steps.ccache_vars.outputs.hash }}-${{ github.ref }}
+            ccache-${{ job.id }}-${{ steps.ccache_vars.outputs.hash }}
+            ccache-${{ job.id }}
```

Comment:
I think I would also put the emscripten version number in the key. There are some benefits to bumping it up (but right now we can't because cmake is not configured correctly, and we're passing emscripten flags in all the wrong places which is an error condition since 2.0.16 I think?)